src/fileformats/ocd_types.cpp

```diff
@@ -100,6 +108,286 @@ namespace Ocd
 	}
 	
 	
+	
+	namespace
+	{
+	
+	/**
+	 * This struct refers to a sequence of bytes in existing data.
+	 */
+	struct ByteSequence {
+		const quint8* start;
+		unsigned int count;
+	};
+	
+	
+	/**
+	 * This class allows to read from and write to streams of 9-bit code words.
+	 * 
+	 * The stream "starts" with the lowest bit, i.e. the lowest bit of the
+	 * first byte of the stream is the lowest bit of the first code word.
+	 * Consequently, the highest bit of the first 9-bit code word is found in
+	 * the lowest bit of the second byte of the stream, and the lowest bit of
+	 * the second code word is found in the second lowest bit of the second byte
+	 * in the stream.
+	 * 
+	 * For writing, the Byte template parameter must not be const.
+	 * 
+	 * Synopsis:
+	 *     using CodeStream = CodeStreamImplementation<quint8>;
+	 *     auto stream = CodeStream{icon.bits};
+	 *     stream.seek(16);
+	 *     stream.put(0);
+	 */
+	template<class Byte>
+	class CodeStreamImplementation
+	{
+		Q_STATIC_ASSERT(sizeof(Byte) == 1);
+		Byte* data;
+		unsigned byte_pos = 0;
+		unsigned bit_pos = 0;
+		unsigned end_pos;
+		
+	public:
+		/**
+		 * Special values in a compressed stream of icon data.
+		 * 
+		 * 0x100u could be "Clear code table". However, this application (9 bit
+		 * code word size, 256 byte input size) shouldn't need it.
+		 */
+		enum SpecialValues
+		{
+			LastStaticCode   = 0x0ffu,
+			Reserved_0x100   = 0x100u,
+			EndOfData        = 0x101u,
+			FirstDynamicCode = 0x102u,
+		};
+		
+		template<class T, unsigned N = std::extent<T>::value>
+		CodeStreamImplementation(T& data) noexcept : data{data}, end_pos(N) { Q_STATIC_ASSERT(N != 0); }
+		
+		bool atEnd() const noexcept { return byte_pos >= end_pos; }
```

Comment:
Hm. This spot should better reflect the fact that put()/get() will access two bytes. put()/get() shall not throw if not atEnd(). With that change, I can even call atEnd() from put()/get().